windows/src/developer/kmanalyze/kmanalyze.cpp

```diff
@@ -0,0 +1,692 @@
+// kmanalyze.cpp : This file contains the 'main' function. Program execution begins and ends there.
+//
+
+#include "pch.h"
+#include "crc32.h"
+#include <iostream>
+#include <vector>
+#include <codecvt>
+#include <locale>
+
+BOOL LoadKeyboard(LPSTR fileName, LPKEYBOARD *lpKeyboard);
+
+BOOL VerifyChecksum(LPBYTE buf, LPDWORD CheckSum, DWORD sz);
+void Err(const char *p);
+int DoKeyboardAnalysis(LPKEYBOARD kbd, char *keyboardID, char *keyboardJSFilename, char *outputfilename);
+void MapVirtualKeys(void);
+
+struct TEST {
+  UINT key;
+  UINT shift;
+  std::wstring context;
+};
+
+typedef std::vector<TEST> TESTS;
+
+enum GROUPREFTYPE { rule, match, nomatch };
+
+struct GROUPTREE;
+
+struct GROUPREF {
+  GROUPREFTYPE type;
+  GROUPTREE* target;
+  LPKEY rule;
+};
+
+struct GROUPTREE {
+  LPGROUP group;
+  std::vector<GROUPREF> refs;
+};
+
+TESTS *DoGroupAnalysis(LPKEYBOARD kbd, LPGROUP gp, std::vector<LPGROUP> & tree, TEST *test);
+
+wchar_t VKToChar[256][2];
+UINT CharToVK[256];
+UINT CharToShift[256];
+
+int main(int argc, char *argv[])
+{
+  LPKEYBOARD kbd;
+  char buf[_MAX_PATH], drive[_MAX_DRIVE], dir[_MAX_DIR], filename[_MAX_FNAME], ext[_MAX_EXT], jsfilename[_MAX_PATH];
+
+  if (argc < 2 || !strcmp(argv[1], "--help"))
+  {
+    puts("KMANALYZE: Extract rules from a Keyman Desktop .kmx keyboard to use for building automated tests");
+    puts("(C) SIL International");
+    puts("Usage: KMANALYZE <filename> [outputfilename]\n");
+    puts("Will create a keyboard.tests from keyboard.js; if outputfilename is not specified,");
```

Comment:
So instead of outputfilename, you can optionally give a list of drive, dir, filename, ext? (line 68)

Reply:
Not really, no. If you don't specify outputfilename, it dumps the .tests file in the same place as the .kmx file (e.g. `foo\build\foo.kmx` would then have `foo\build\foo.tests`. The regression test suite in #1682 does specify the output filename though...